keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+importing kubernetes/kubernetes.
+
+Currently importing the k/k root module will make your build fail.
+The best solution for users is to fork the kubeadm API packages.
+Once kubeadm moves out, importing the "kubeadm" module will make
+it possible to easily import only the API portion.
+
+#### Story 3
+
+As a user, I would like to import utility packages from kubeadm.
+
+Similar to the above user story, importing kubeadm from k/k will fail.
+Once kubeadm moves out, utility functions will be exposed for public use,
+but perhaps without any guarantees until the kubeadm backend is truly
+announced for public consumption.
+
+### Implementation Details/Notes/Constraints
+
+#### Location
+
+The proposed new location is the kubernetes/kubeadm repository.
+
+This makes it possible to use `go get k8s.io/kubeadm` to obtain
+the kubeadm source tree.
+
+The current contents of the `k/k/cmd/kubeadm/app` directory
+should be moved in the root of k/kubeadm repository.
+
+The current `k/k/cmd/kubeadm/test` directory that contains
+integration tests should be merged with the `k/kubeadm/tests`
+directory under `k/kubeadm/tests/integration`.
+
+Existing tools in the k/kubeadm repository such as "kinder" and
+"operator" must continue existing as separate modules (sub-folders)
+inside the kubeadm module and can be obtained using
+`go get k8s.io/kubeadm/some-tool`.
+
+Existing docs and tests must co-exist in sub-folders too.
+
+Side effects:
+- All contents of k/kubeadm will be branched and tagged with kubeadm,
+but this is actually a positive change.
+- `hack/verify*` scripts for the root kubeadm folder should skip "kinder"
+and "operator". This area can see improvements in the future,
+potentially sharing scripts for all modules.
+
+#### Handling older kubeadm versions
+
+Currently kubeadm in the `release-1.17` branch (and later) of k/k
+does not depend on the rest of k/k. This was required for kubeadm
+to move out, however the older branches of k/k include versions of
+kubeadm that still depend on k/k.
+
+The proposed way forward is to move only the latest kubeadm from k/k master
+and build older version of kubeadm from k/k.
+
+This creates the following complications:
+- The post-submit job must handle tags and branches for kubeadm versions
+larger than X.
+- The kubernetes/release tooling must build kubeadm version X from the
+kubeadm repository while build older kubeadm from k/k.
+
+The cherry-pick script from k/k/hack will not work out of the box
+for the kubeadm repository so potentially it can be forked to suite
+kubeadm. The same cherry pick policy will apply as in k/k.
+
+#### Releasing and tagging
+
+##### Challenges
+
+The current plan is for kubeadm to continue to follow the k8s version scheme
+and release schedule. This will continue to make the skew of kubeadm
+and k8s easy to understand for the users and continue to give motivation
+to the kubeadm developers to get work done on time.
```

Comment:
When people talk about moving out of tree, the two most frequently cited benefits I hear are:

 * being able to run at an independent release cadence (faster or slower)
 * independent semantic versioning (move to 2.0 when you have a major change)

While avoiding those two benefits this KEP feels like it manufactures a lot of tagging and synchronization complexity.  Why not take the plunge aim to achieve those two benefits as a strong stand-alone project?  In such a scenario would it be possible for kubeadm to itself track and report to users what versions and skew of components are supported by a given kubeadm version, so as to still keep things clear for them?

Reply:
> Why not take the plunge aim to achieve those two benefits as a strong stand-alone project?
> In such a scenario would it be possible for kubeadm to itself track and report to users what versions and skew of components are supported by a given kubeadm version, so as to still keep things clear for them?

- this will break thousands of people and their infrastructure.
- people will stop understanding what version of kubeadm supports which versions of k/k without going online and looking at a version matrix.

following the k8s versioning was something that arguably saw the most discussion and i don't think the kubeadm maintainers will change their mind at this point.

> While avoiding those two benefits this KEP feels like it manufactures a lot of tagging and synchronization complexity

this is true, but what happens when one day the kubelet and the control-plane components move out? should they also follow their unique sem.vers?

also k8s already has synchronization and automation via the publishing bot. but it cannot be used in this case.

Reply:
> this is true, but what happens when one day the kubelet and the control-plane components move out? should they also follow their unique sem.vers?

I've presumed they would and the "release" at that point truly would be declaring the content package names-versions-releases for the distribution.

Reply:
> following the k8s versioning was something that arguably saw the most discussion and i don't think the kubeadm maintainers will change their mind at this point.

I don't see a fundamental conflict there.  But I also don't think that following or not following the k/k version solves the question of which k/kubeadm content is ready for release in conjunction with k/k and ensures that k/k is never released with the wrong k/kubeadm content.

Reply:
I'm primarily concerned that the KEP mostly focuses on a prow based auto branch/tag scheme.

If CI were always green and we never had PRs in a repo or across repos that depended on each other, I'd be more confident in the described approach.  In the meantime I foresee us regularly deleting tags and retagging at the "right" place when automation was hasty, or us burning through release numbers without making an actual release in order to get to a point where the vX.YY.Z tags on the various repos are all at the same number and include all the desired commits and no-non-desired commits.

Reply:
> I don't see a fundamental conflict there. But I also don't think that following or not following the k/k version solves the question of which k/kubeadm content is ready for release in conjunction with k/k and ensures that k/k is never released with the wrong k/kubeadm content.

the Qt are handling that pretty well.
i'm not aware who performs the tagging of their sub-modules but i don't think it's that important, as a human mistake can still include a bogus commit in a release.

mistakes happen. i don't think this KEP should be answering the fundamental question of "is a k8s component ready for a release".

Reply:
This may be something for us to bubble up to an arch level conversation.  For the near term, it isn't needing resolved for this KEP.